code/localization/localize.cpp

```diff
@@ -724,199 +717,321 @@ void lcl_fred_replace_stuff(SCP_string &text)
+		if (strnicmp(ch, "XSTR", 4) != 0)
+			break;
+		ch += 4;
 
-		if (id != NULL)
-			*id = -2;
+		attempted_xstr = true;
 
-		return false;
-	}
+		// the next non-whitespace char should be a (
+		ignore_white_space(&ch);
+		if (*ch != '(')
+			break;
+		ch++;
 
-	// otherwise, check to see if it's an XSTR() tag
-	if (strnicmp(in.c_str(), "XSTR", 4) != 0) {
-		// NOT an XSTR() tag
-		out = in;
+		// the next should be a quote
+		ignore_white_space(&ch);
+		if (*ch != '\"')
+			break;
+		ch++;
 
-		if (id != NULL)
-			*id = -2;
+		// now we have the start of the string
+		auto str_start = ch;
 
-		return false;
-	}
+		// find the end of the string
+		ch = strchr(ch, '"');
+		if (ch == nullptr)
+			break;
 
-	// at this point we _know_ its an XSTR() tag, so split off the strings and id sections		
-	if (!lcl_ext_get_text(in, text_str)) {
-		out = in;
+		// now we have the end of the string (past the last character in it)
+		auto str_end = ch;
+		ch++;	// skip the quote
 
-		if (id != NULL)
-			*id = -1;
+		// the next non-whitespace char should be a ,
+		ignore_white_space(&ch);
+		if (*ch != ',')
+			break;
+		ch++;
 
-		return false;
-	}
-	if (!lcl_ext_get_id(in, &str_id)) {
-		out = in;
+		// check for number, being mindful of negative
+		ignore_white_space(&ch);
+		bool is_negative = false;
+		if (*ch == '-')
+		{
+			is_negative = true;
+			ch++;
+		}
+		if (!isdigit(*ch))
+			break;
 
-		if (id != NULL)
-			*id = -1;
+		// now we have the start of the id
+		auto id_start = ch;
 
-		return false;
-	}
-	
-	// if the localization file is not open, or there's no entry, or we're not translating, return the original string
-	if ( !Xstr_inited || (str_id < 0) || (!use_default_translation && ((Lcl_current_lang == LCL_UNTRANSLATED) || (Lcl_current_lang == LCL_RETAIL_HYBRID))) ) {
-		out = text_str;
+		// find all the digits
+		while (isdigit(*ch))
+			ch++;
 
-		if (id != NULL)
-			*id = str_id;
+		// now we have the end of the id (past the last character in it)
+		auto id_end = ch;
 
-		return true;
-	}
+		// the next non-whitespace char should be a )
+		ignore_white_space(&ch);
+		if (*ch != ')')
+			break;
 
-	auto lookup_map = &Lcl_ext_str;
-	if (use_default_translation && lcl_get_current_lang_index() != LCL_DEFAULT) {
-		// if we're not already using the default, then switch to our explicit default
-		lookup_map = &Lcl_ext_str_explicit_default;
-	}
+		// if we got this far, we know we have a parseable XSTR of some sort
+		xstr_id = -1;
 
-	// get the string if it exists
-	if (lookup_map->find(str_id) != lookup_map->end()) {
-		// copy to the outgoing string
-		out = (*lookup_map)[str_id];
-	}
-	// otherwise use what we have - probably should Int3() or assert here
-	else {
-		out = text_str;
-	}
+		//
+		// split off the strings and id sections
+		//
 
-	// set the id #
-	if (id != NULL){
-		*id = str_id;
-	}
+		// now that we know the boundaries of the actual string in the XSTR() tag, copy it
+		text_str.assign(str_start, str_end);
```

Comment:
Shouldn't the second argument be `str_end - str_start`?

Reply:
I haven't investigated whether that would work, but this also works.  This treats the function parameters as a start iterator and an end iterator.